pkg/controller/utils/utils.go

```diff
@@ -151,9 +152,21 @@ func IsCrashed(pod *v1.Pod) bool {
 
 // IsImageEqual returns true if image name image1 is equal to image name image2.
 func IsImageEqual(image1 string, image2 string) bool {
-	desiredImage := strings.TrimPrefix(image1, DockerHubImagePrefix)
-	actualImage := strings.TrimPrefix(image2, DockerHubImagePrefix)
-	return actualImage == desiredImage
+	desiredImageWithVersion := strings.TrimPrefix(image1, DockerHubImagePrefix)
+	actualImageWithVersion := strings.TrimPrefix(image2, DockerHubImagePrefix)
+
+	desiredRegistry, desiredName, desiredVersion := ParseDockerImageTag(desiredImageWithVersion)
+	actualRegistry, actualName, actualVersion := ParseDockerImageTag(actualImageWithVersion)
+
+	// registry name, image name and version should match.
+	return desiredRegistry == actualRegistry && desiredName == actualName && (desiredVersion == actualVersion || (desiredVersion == ":latest" && actualVersion == "") || (actualVersion == ":latest" && desiredVersion == ""))
```

Comment:
Is it just a defensive check to see if desiredVersion is :latest or it can be this? I think you have added a check to disallow user from providing empty or :latest version.

Reply:
Yes this is intentional because if user specifies ":latest" or empty then comparing versions makes no sense. For the latest, the operator will have to check the latest version and we could unintentionally upgrade if we push new versions to the docker hub.